### PR TITLE
rd-cpp: miscellaneous compilation fixes for newer compilers and c++ standarts

### PR DESCRIPTION
**Pull request by LeoMash**
* Ported some fixes from UnrealLink
* Updated gtest & spdlog dependencies
* Fixed several deprecated STL usages (still leaving compatibility with C++14)
* Fixed compilation warnings on clang

Now rd-cpp should successfully compile on both clang, msvc and gcc (with some warnings left) in c++14, c++17 and c++20 standarts

### CODE REVIEW

File: rd-cpp/src/rd_framework_cpp/src/main/protocol/Buffer.cpp [hunk 1]
```diff
@@ -190,7 +205,7 @@ int64_t TICKS_PER_MILLISECOND = 10000000;
 DateTime Buffer::read_date_time()
 {
 	int64_t time_in_ticks = read_integral<int64_t>();
-	time_t t = static_cast<time_t>(time_in_ticks / TICKS_PER_MILLISECOND - TICKS_AT_EPOCH / TICKS_PER_MILLISECOND);
+	time_t t = static_cast<time_t>((time_in_ticks - TICKS_AT_EPOCH) / TICKS_PER_MILLISECOND);
```
Comment (ForNeVeR) [inline rd-cpp/src/rd_framework_cpp/src/main/protocol/Buffer.cpp:208]: Here, you're rolling back a change purposefully made at https://github.com/JetBrains/rd/commit/4be1d02063cfaff0781b7ae23a1822142505a8ee.

Are you sure this should be done? Perhaps we should fix `write_date_time` instead, too?

File: rd-cpp/src/rd_framework_cpp/src/main/serialization/Polymorphic.cpp [hunk 2]
```diff
@@ -2,8 +2,4 @@
 
 namespace rd
 {
-template class Polymorphic<int8_t>;
-template class Polymorphic<int16_t>;
-template class Polymorphic<int32_t>;
-template class Polymorphic<int64_t>;
-}	 // namespace rd
\ No newline at end of file
+}	 // namespace rd
```
Comment (ForNeVeR) [inline rd-cpp/src/rd_framework_cpp/src/main/serialization/Polymorphic.cpp:5]: Do we need this file now at all?

File: rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp [hunk 3]
```diff
@@ -146,7 +153,7 @@ void SocketWire::Base::set_socket_provider(std::shared_ptr<CActiveSocket> new_so
 	});
 	const auto status = heartbeat.wait_for(timeout);
 
-	logger->debug("{}: waited for heartbeat to stop with status: {}", this->id, status);
+	logger->debug("{}: waited for heartbeat to stop with status: {}", this->id, (uint32_t)status);
```
Comment (ForNeVeR) [inline rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp:156]: _(Minor)_ `static_cast`, maybe?

File: rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp [hunk 4]
```diff
@@ -400,7 +407,7 @@ void SocketWire::Base::ping() const
 	}
 	catch (std::exception const& e)
 	{
-		logger->debug("{}: exception raised during PING | {}", this->id, e.what());
+		logger->warn("{}: exception raised during PING | {}", this->id, e.what());
```
Comment (ForNeVeR) [inline rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp:410]: People were asking us to stop logging ping errors to `warn`. Please roll this back; it was done on purpose.

File: rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp [hunk 5]
```diff
@@ -506,7 +509,7 @@ SocketWire::Client::Client(Lifetime lifetime, IScheduler* scheduler, uint16_t po
 		{
 			logger->info("{}: closed with exception: {}", this->id, e.what());
 		}
-		logger->info("{}: terminated, port: {}.", this->id, this->port);
+		logger->debug("{}: thread expired", this->id);
```
Comment (ForNeVeR) [inline rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp:512]: Please put back the logging statements in this function.

File: rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp [hunk 6]
```diff
@@ -552,55 +561,53 @@ SocketWire::Server::Server(Lifetime lifetime, IScheduler* scheduler, uint16_t po
 	RD_ASSERT_MSG(this->port != 0, fmt::format("{}: port wasn't chosen", this->id));
 
 	logger->info("{}: listening 127.0.0.1/{}", this->id, this->port);
+	Lifetime lifetime = serverLifetimeDefinition.lifetime;
 
 	thread = std::thread([this, lifetime]() mutable {
 		rd::util::set_thread_name(this->id.empty() ? "SocketWire::Server Thread" : this->id.c_str());
 
-		logger->info("{}: started, port: {}.", this->id, this->port);
```
Comment (ForNeVeR) [inline rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp:559]: Please revert this logging statement back.

File: rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp [hunk 7]
```diff
@@ -552,55 +561,53 @@ SocketWire::Server::Server(Lifetime lifetime, IScheduler* scheduler, uint16_t po
...
+			{
+				logger->info("{}: closed with exception: {}", this->id, e.what());
 			}
 		}
-		catch (std::exception const& e)
-		{
-			logger->error("{}: terminal socket error ({}).", this->id, e.what());
-		}
-
-		logger->info("{}: terminated, port: {}.", this->id, this->port);
```
Comment (ForNeVeR) [inline rd-cpp/src/rd_framework_cpp/src/main/wire/SocketWire.cpp:603]: Let's have this "terminated+port" logging statement, and delete the unclear one `logger->debug("{}: thread expired", this->id)`.

**Review by ForNeVeR — CHANGES_REQUESTED** [review]

File: rd-cpp/src/rd_framework_cpp/src/main/protocol/Buffer.cpp (hunk 1 above)
Comment (LeoMash) [reply rd-cpp/src/rd_framework_cpp/src/main/protocol/Buffer.cpp:208]: Sorry, I think I messed up when ported changes from UnrealLink. I though that these were made there, and no changes were in rd-cpp repo. I will reevaluate changes

Comment (ForNeVeR) [reply rd-cpp/src/rd_framework_cpp/src/main/protocol/Buffer.cpp:208]: I think I've marked every case when something wrong was happening. Everything else LGTM.

Comment (LeoMash) [reply rd-cpp/src/rd_framework_cpp/src/main/protocol/Buffer.cpp:208]: Thank you!

**Review by ForNeVeR — APPROVED** [review]
LGTM.

**Pull request merged**